### ISSUE TITLE
Proper handling of "waiting for service" in integration tests

### ISSUE THREAD

**Issue by ChrsMark** [elastic/beats#17177]
In cases like `Prometheus` Metricbeat, we may need to retry so as to have metrics to be populated, like in [this](https://github.com/elastic/beats/blob/master/metricbeat/module/prometheus/query/query_integration_test.go#L84) integration test.

With @sayden we have been wondering if this could lead in potential flakiness. However we do something in similar in `system-tests` where we call [`wait_until`](https://github.com/elastic/beats/blob/450067d343c0702819c293eb67078b12c2a3ff40/libbeat/tests/system/beat/beat.py#L352) with timeouts.

The purpose of this issue is to discuss and conclude if this is something that is acceptable or it is something that can be used in golang integration-tests of the project. 

Also if we reach out to have this accepted we may need to implement this retry approach in a helper function for reusability.

cc: @elastic/integrations-platforms @elastic/integrations-services

**Comment by exekias** [comment]
I see a point for a helper method to wait for a condition to happen, very similar to what we have in python tests. As usual, it needs to have a sane timeout by default (ie 10 seconds).

In the long run (unrelated to this issue) I would prefer to bring the scenario up in a previous step, and then run tests on top of it. It means we shift the responsibility of bringing up a healthy scenario outside of the testing code, and simplify tests to suppose that the scenario is already there.

That would also allow us to launch tests faster when developing, as you can leave the scenario running and keep running your tests against it

**Comment by exekias** [comment]
pinging @jsoriano as he has been thinking about this for a while

**Comment by jsoriano** [comment]
TLDR; This could be implemented as an `UpOption` of `EnsureUp(service, opts...)`. But I think that this wouldn't be the way to go, so I don't think that we should facilitate waits in the integration tests framework.

I see two scenarios where this could be needed:
* The service needs some warm-up time: then the wait should be part of the health-check, not part of the test.
* The module needs to query multiple times, because it calculates something from previous values, or to do something because this is what actually warms-up the stats in the service: then I think that these previous queries should be part of the test, or the metricset, as they would be quite specific to the service.
  On these cases we also have to be extra careful, in case we make some query in tests that we don't do in the real metricset and then the real metricset may not be working.

> However we do something in similar in `system-tests` where we call [`wait_until`](https://github.com/elastic/beats/blob/450067d343c0702819c293eb67078b12c2a3ff40/libbeat/tests/system/beat/beat.py#L352) with timeouts.

Take into account that this is not the same. System tests are black box tests, where we start the beat in background and we don't know when things are going to happen. In these tests there are some cases where we want to check that something has appeared in logs, or that some events have been written to the file, all these things happen asynchronously, and the test doesn't have much control on them, thus the waits. In integration tests we have (or we should have) full control on when things happen.

>In the long run (unrelated to this issue) I would prefer to bring the scenario up in a previous step, and then run tests on top of it. It means we shift the responsibility of bringing up a healthy scenario outside of the testing code, and simplify tests to suppose that the scenario is already there.

Adding better health checks, or improving the docker image would be in this line. Adding options for the testing code would be in the opposite direction.

> That would also allow us to launch tests faster when developing, as you can leave the scenario running and keep running your tests against it

Integration tests currently leave the services running, so they are reused. In most cases in system and integration tests the service host can be specified with `<service>_HOST` environment variable too and then compose is not used, what allows to reuse scenarios, or query any endpoint during tests. Though this is something to improve, because it doesn't work in all cases, and it should be also better documented.
